### PR TITLE
Generate a script to handle coverage target at build time

### PR DESCRIPTION
**Pull request by isaachier**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -239,15 +239,15 @@ endif()
 find_package(Doxygen)
 option(ENABLE_DOCS "Enable building documentation." ON)
 if(DOXYGEN_FOUND AND ENABLE_DOCS)
-    set_property(DIRECTORY APPEND PROPERTY ADDITIONAL_MAKE_CLEAN_FILES "${PROJECT_BINARY_DIR}/dev-docs/_build")
+    set_property(DIRECTORY APPEND PROPERTY ADDITIONAL_MAKE_CLEAN_FILES "${CMAKE_CURRENT_BINARY_DIR}/dev-docs/_build")
```
Comment (isaacbrodsky) [inline CMakeLists.txt]: It looks like this change would help if H3 were included as a subproject, is that right?

File: cmake/RunCoverage.cmake.in [hunk 2]
```diff
@@ -0,0 +1,15 @@
+if($<$<CONFIG:Debug>:1>)
```
Comment (isaacbrodsky) [inline cmake/RunCoverage.cmake.in]: Please add copyright header.

File: cmake/RunCoverage.cmake.in [hunk 3]
```diff
@@ -0,0 +1,15 @@
+if($<BOOL:$<CONFIG:Debug>>)
```
Comment (isaacbrodsky) [inline cmake/RunCoverage.cmake.in]: Please add copyright header & an explanation of the purpose of this file.

File: CMakeLists.txt (hunk 1 above)
Comment (isaachier) [reply CMakeLists.txt]: Yes. Just always a good idea to use `CMAKE_CURRENT_SOURCE_DIR`/`CMAKE_CURRENT_BINARY_DIR` as opposed to their `PROJECT_...` counterparts.

File: cmake/RunCoverage.cmake.in (hunk 3 above)
Comment (isaachier) [reply cmake/RunCoverage.cmake.in]: Sure thing. Good points.

**Comment by isaacbrodsky** [comment]
Looks good to me, but it looks like AppVeyor and Coveralls are having issues with this change

**Comment by isaachier** [comment]
Good point. I know why this is happening. Just a second.

**Comment by isaacbrodsky** [comment]
Looks like `Debug/` and `Release/` need to be gitignore'd

I wasn't able to run coverage. I did:
```
git clean #...
cmake .
make coverage
# fails: expected
cmake -DCMAKE_BUILD_TYPE=Debug .
make coverage
# fails: unexpected
```

The error is like:
```
[100%] Built target testH3ToGeoBoundary_coverage51
Capturing coverage data from <...>/Debug
Found LLVM gcov version 9.1.0, which emulates gcov version 4.2.0
Scanning <...>/Debug for .gcda files ...
geninfo: WARNING: no .gcda files found in <...>/Debug - skipping!
Finished .info-file creation
lcov: ERROR: no valid records found in tracefile coverage.info
Reading tracefile coverage.info
genhtml: ERROR: cannot read file coverage.cleaned.info!
Reading data file coverage.cleaned.info
-- Coverage done
[100%] Built target coverage
```

It looks like CI also had this issue: https://travis-ci.org/uber/h3/jobs/418413706

**Comment by isaachier** [comment]
Another good catch. I believe all my testing relied on the build being in a directory of h3.

File: cmake/RunCoverage.cmake.in [hunk 4]
```diff
@@ -31,16 +31,20 @@
 # ```
 
 if($<BOOL:$<CONFIG:Debug>>)
-    get_filename_component(root_dir "${CMAKE_CURRENT_LIST_DIR}" DIRECTORY)
+    set(source_dir "${CMAKE_ARGV0}")
```
Comment (isaacbrodsky) [inline cmake/RunCoverage.cmake.in]: Might want to add a comment at the top of the script that it expects the source_dir and binary_dir as arguments in that order.

**Comment by isaacbrodsky** [comment]
I think the coverage CI test (`.travis.yml` around line 57) might still need updating, since Coveralls doesn't show the coverage for this PR. https://travis-ci.org/uber/h3/jobs/418460350

**Comment by isaacbrodsky** [comment]
I'm not getting the coverage information generated when testing this branch:

On 3f99a6 after running `cmake -DCMAKE_BUILD_TYPE=Debug ..` and then:
```
$ make coverage
...
[ 98%] Built target testLinkedGeo_coverage91
Scanning dependencies of target testH3ToGeo_coverage3
[100%] Running testH3ToGeo_coverage3
[100%] Built target testH3ToGeo_coverage3
Scanning dependencies of target testH3ToGeoBoundary_coverage51
[100%] Running testH3ToGeoBoundary_coverage51
[100%] Built target testH3ToGeoBoundary_coverage51
Scanning dependencies of target coverage
-- Coverage done
[100%] Built target coverage
$ ls Debug
RunCoverage.cmake
```

**Comment by isaachier** [comment]
@isaacbrodsky what do you expect to see? It is executing a subprocess and might not be redirecting the output correctly.

**Comment by isaacbrodsky** [comment]
We should probably redirect the output, so we can see the coverage. I can't find for example the `index.html` file `genhtml` should be generating.

Similarly it looks like CI is still failing: https://travis-ci.org/uber/h3/jobs/418489867 `IOError: [Errno 2] No such file or directory: 'Debug/coverage.cleaned.info'`

File: cmake/coverage.ps1.in [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+If ($<BOOL:$<CONFIG:Debug>> -eq 0) {
+    Write-Host "Cannot run coverage for non-debug build"
+    Exit
+}
+
+$src_dir=${1:-"Missing source directory"}
+binary_dir=${2:-"Missing binary directory"}
+
+cd "${binary_dir}"
+lcov --directory . --capture --output-file coverage.info
```
Comment (isaacbrodsky) [inline cmake/coverage.ps1.in]: Is there lcov for Windows? From what I saw it looked like no.

File: cmake/coverage.sh.in [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    exit 1
+fi
+
+src_dir=${1:-"Missing source directory"}
+binary_dir=${2:-"Missing binary directory"}
+
+cd "${binary_dir}"
+lcov --directory . --capture --output-file coverage.info
+lcov --extract coverage.info "${src_dir}/src/h3lib/*" --output-file coverage.cleaned.info
+genhtml -o coverage coverage.cleaned.info --title 'h3 coverage'
```
[the remark this answers is not in the document]
Comment (isaacbrodsky) [reply cmake/coverage.sh.in]: `genhtml` is used in master.

File: CMakeLists.txt [hunk 7]
```diff
@@ -330,21 +330,16 @@ if(BUILD_TESTING)
...
         add_custom_target(coverage
-            # Compile generated coverage counts into a single coverage.info
-            COMMAND lcov --directory '${PROJECT_BINARY_DIR}' --capture --output-file coverage.info
-            # Take only coverage of the core library, not apps
-            COMMAND lcov --extract coverage.info '${PROJECT_SOURCE_DIR}/src/h3lib/*' --output-file coverage.cleaned.info
-            # Generate output html at coverage/
-            COMMAND genhtml -o coverage coverage.cleaned.info --title 'h3 coverage'
-            COMMENT "Coverage done"
-            )
+            COMMAND bash "${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/coverage.sh" "${CMAKE_CURRENT_SOURCE_DIR}" "${CMAKE_CURRENT_BINARY_DIR}")
```
[the remark this answers is not in the document]
Comment (isaachier) [reply CMakeLists.txt]: I originally did but @isaacbrodsky points out that Windows probably doesn't support lcov without bash anyway (i.e. Cygwin), so coverage on Windows hasn't ever been supported.

File: cmake/coverage.sh.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (isaachier) [reply cmake/coverage.sh.in]: If you care to make the error a bit more clear, I can set up a `find_program` command with better error message.

**Pull request merged**